src/main/java/org/broadinstitute/hellbender/tools/walkers/bqsr/BaseRecalibratorUprooted.java

```diff
@@ -0,0 +1,564 @@
+     */
+    private Covariate[] requestedCovariates;
+
+    private RecalibrationEngine recalibrationEngine;
+
+    private int minimumQToUse;
+
+    private static final String NO_DBSNP_EXCEPTION = "This calculation is critically dependent on being able to skip over known variant sites. Please provide a VCF file containing known sites of genetic variation.";
+
+    private BAQ baq; // BAQ the reads on the fly to generate the alignment uncertainty vector
+    private ReferenceDataSource referenceDataSource; // datasource for the reference. We're using a different one from the engine itself to avoid messing with its caches.
+    private final static byte NO_BAQ_UNCERTAINTY = (byte)'@';
+
+
+    private long accumulator;
+
+    public BaseRecalibratorUprooted(SAMFileHeader samHeader, RecalibrationArgumentCollection toolArgs) {
+        this.RAC = toolArgs;
+        this.samHeader = samHeader;
+    }
+
+    @Override
+    public boolean requiresReference() {
+        return true;
+    }
+
+
+    /**
+     * Parse the -cov arguments and create a list of covariates to be used here
+     * Based on the covariates' estimates for initial capacity allocate the data hashmap
+     */
+    @Override
+    public void onTraversalStart() {
+        throw new UnsupportedOperationException("not implemented");
+    }
+
+    public void onTraversalStart(File refFile) {
+        accumulator= 0L;
+
+        baq = new BAQ(BAQGOP); // setup the BAQ object with the provided gap open penalty
+
+        if (RAC.FORCE_PLATFORM != null)
+            RAC.DEFAULT_PLATFORM = RAC.FORCE_PLATFORM;
+
+        requestedCovariates = getCovariatesArray();
+
+        logger.info("The covariates being used here: ");
+        for (Covariate cov : requestedCovariates) { // list all the covariates being used
+            logger.info("\t" + cov.getClass().getSimpleName());
+            cov.initialize(RAC); // initialize any covariate member variables using the shared argument collection
+        }
+
+        // no output file. We want to be able to output to a PCollection.
+
+        initializeRecalibrationEngine();
+        minimumQToUse = PRESERVE_QSCORES_LESS_THAN;
+
+        referenceDataSource = new ReferenceDataSource(refFile);
+    }
+
+    private Covariate[] getCovariatesArray() {
+        Pair<ArrayList<Covariate>, ArrayList<Covariate>> covariates = RecalUtils.initializeCovariates(RAC); // initialize the required and optional covariates
+        List<Covariate> covariatesList = new ArrayList<>(covariates.getLeft().size() + covariates.getRight().size());
+        covariatesList.addAll(covariates.getLeft());
+        covariatesList.addAll(covariates.getRight());
+        return covariatesList.toArray(new Covariate[covariatesList.size()]);
+    }
+
+    @Override
+    public SAMFileHeader getHeaderForReads() {
+        return samHeader;
+    }
+
+    /**
+     * Initialize the recalibration engine
+     */
+    private void initializeRecalibrationEngine() {
```

Comment:
This seems a little weird to me.

This is a two-line function that is only called from one other function and (implicitly) assumes that a package-level variable has already been constructed. It would make more sense to me get rid of this as a function (move it up into onTraversalStart) unless you are going to unit test it by itself, which case I think it should take requestedCovariates as an argument.